### PR TITLE
sympify : passing `evaluate=False` flag to `Basic` class methods

### LINKED ISSUE #11441: parse_expr(u'Abs(x+x)' , evaluate=False) reduces to 2*Abs(x).  For me this is very bad
I'm dealing with significant digits, so I NEVER want the expression to reduce itself.  Using parse_expr is generally what I need, but when I throw Abs into the mix, it goes haywire.

This is also the case when I attempt to substitute my own abs function using the global_dict.

### LINKED ISSUE #13999: sympify should pass evaluate=False to SymPy functions
The flag evaluate=False of `sympify` is meant to prevent evaluation: 
```
>>> sympify("2.2 + 3.3", evaluate=False)
2.2 + 3.3
```
But it doesn't help with functions of a floating point argument:
```
>>> sympify("log(5.5)", evaluate=False)
1.70474809223843
```
This is despite `log`, and other SymPy functions, supporting `evaluate=False` on their own: 
```
>>> log(5.5, evaluate=False)
log(5.5)
```

If ` evaluate=False` is set in `sympify`, it should propagate to any SymPy functions found in the parsed string.

### PR DESCRIPTION
**Pull request by jashan498**
Fixes #13999
Added `visit_call` method inside `EvaluateFalseTransformer` to stop evaluation of functions which inherits from `Basic`.
```
>>> sympify("log(5.5)", evaluate=False)
log(5.5)
>>> sympify("log(5 - 3)", evaluate=False)
log(-3 + 5)
>>> sympify("Mul(3, 5.5)", evaluate=False)
3*5.5
>>> sympify("log(log(5 - 3))", evaluate=False)
log(log(-3 + 5))
```
Also fixes  #11441
```
>>> parse_expr("Abs(x+x)", evaluate=False) 
Abs(x + x)
```

### CODE REVIEW

**Comment by asmeurer** [comment]
I got this error 

```pytb
>>> sympify('1 + 1 + log(2 + 2)', evaluate=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./sympy/core/sympify.py", line 368, in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
  File "./sympy/parsing/sympy_parser.py", line 948, in parse_expr
    code = compile(evaluateFalse(code), '<string>', 'eval')
  File "./sympy/parsing/sympy_parser.py", line 958, in evaluateFalse
    node = EvaluateFalseTransformer().visit(node)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/ast.py", line 300, in generic_visit
    value = self.visit(value)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/ast.py", line 309, in generic_visit
    new_node = self.visit(old_value)
  File "/Users/aaronmeurer/anaconda3/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "./sympy/parsing/sympy_parser.py", line 1049, in visit_BinOp
    new_node.args = self.flatten(new_node.args, sympy_class)
  File "./sympy/parsing/sympy_parser.py", line 980, in flatten
    if isinstance(arg, ast.Call) and arg.func.id == func:
AttributeError: 'Attribute' object has no attribute 'id'
>>>
```

File: sympy/parsing/sympy_parser.py [hunk 1]
```diff
@@ -983,6 +983,30 @@ def flatten(self, args, func):
...
         return result
 
+    def visit_Call(self, node):
+        from sympy import sympify
+        from sympy.core.expr import Expr, AtomicExpr
+        node_class = sympify(node.func.id)
+        if isinstance(node_class, type) and issubclass(node_class, Basic) and not issubclass(node_class, AtomicExpr):
+            args = []
+            for arg in node.args:
+                if isinstance(arg, ast.BinOp):
```
Comment (asmeurer) [inline sympy/parsing/sympy_parser.py]: The docs say one should use `generic_visit` to handle subnodes. Also, does the visit_BinOp do this correctly?

Comment (jashan498) [reply sympy/parsing/sympy_parser.py]: using
```
if isinstance(arg, (ast.BinOp, ast.Call)):
                    args.append(self.generic_visit(arg))
```
doesnot stop evaluation of subnodes.

**Comment by jashan498** [comment]
The above method fails in a lot of cases involving `Pow`, `Mul` and `Add`.

How about using "wrapping alternatives"  you mentioned?  We could have classes like [UnevaluatedExpr](https://github.com/sympy/sympy/blob/eea43ed1983d224b4977abc08fe9e378487411dc/sympy/core/expr.py#L3426) with all the required assumptions initialized and use them for our purpose.

**Comment by jashan498** [comment]
```
>>> This method fails in a lot of cases involving Pow, Mul and Add
```
Evaluation of these can be stopped by using there corresponding operators(like `sympify("3**5.5",evaluate=False)=3**5.5`)), so can we just avoid touching these types inside `visit_call` ?

**Comment by asmeurer** [comment]
That's a good idea. 

I think in general it ought to pass the `evaluate=False` keyword argument if it is supported, and only use the `Basic.__new__` trick if it isn't. But this is not something that can be introspected very easily. 

Objects need some way to signal that they support `evaluate=False`. A simple way would be an attribute on the object. It would be better if instead of `evaluate=False` a special constructor were used, with the default constructor implemented on Basic. 

By the way, these are just ideas. I'm not saying that they should be implemented on this PR. I should probably start a wiki page with my thoughts on unevaluated expressions.

**Comment by czgdp1807** [comment]
@jashan498 Are you still interested to work on this PR?

**Comment by jashan498** [comment]
OH, completely forgot about this PR. Closing

**Pull request closed without merging**